In [16]:
 from googleapiclient import discovery
import json

API_KEY = 'AIzaSyDGTZtyRS-7U1DgN_hMlMn1orwgvJtELSM'

client = discovery.build(
  "commentanalyzer",
  "v1alpha1",
  developerKey=API_KEY,
  discoveryServiceUrl="https://commentanalyzer.googleapis.com/$discovery/rest?version=v1alpha1",
  static_discovery=False,
)

analyze_request = {
  'comment': { 'text': 'friendly greetings from python' },
  'requestedAttributes': {'TOXICITY': {}}
}

response = client.comments().analyze(body=analyze_request).execute()
print(json.dumps(response, indent=2))

{
  "attributeScores": {
    "TOXICITY": {
      "spanScores": [
        {
          "begin": 0,
          "end": 30,
          "score": {
            "value": 0.02543884,
            "type": "PROBABILITY"
          }
        }
      ],
      "summaryScore": {
        "value": 0.02543884,
        "type": "PROBABILITY"
      }
    }
  },
  "languages": [
    "en"
  ],
  "detectedLanguages": [
    "en"
  ]
}


In [17]:
toxic_base = 0.5
threats_and_insults = ['I will kill you!', 'You\'re a complete idiot!', 'Kill yourself', 'Freak me', 'You\'re gay', 'Hideous!', 'You\'re fat!', 'I hope you trip and die', 'Go away!', 'Your moms a hoe!', 'You\'re better off dead!!', 'No one loves you!', 'SHUT UP DUMBASS', 'I hope you fall', 'F uk your politeness c 0k su ker', 'WTF F YOU BASTARDS', 'CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT CUNT', 'You are boring', 'Damn you bastard you killed me DEADGUY', 'Baylor kids should execute themselves', 'Execute yourself!!', 'You\'re so ugly' ]

for text in threats_and_insults:
    analyze_request = {
        'comment': {'text': text},
        'requestedAttributes': {'TOXICITY': {}}
    }

    response = client.comments().analyze(body=analyze_request).execute()
    toxicity_score = response['attributeScores']['TOXICITY']['summaryScore']['value']

    if toxicity_score > toxic_base:
        print(f'"{text}" IS a toxic comment with a toxicity score of {toxicity_score:.2f}')
    else:
        print(f'"{text}" is NOT a toxic comment with a toxicity score of {toxicity_score:.2f}')


"I will kill you!" IS a toxic comment with a toxicity score of 0.93
"You're a complete idiot!" IS a toxic comment with a toxicity score of 0.92
"Kill yourself" IS a toxic comment with a toxicity score of 0.85
"Freak me" is NOT a toxic comment with a toxicity score of 0.36
"You're gay" IS a toxic comment with a toxicity score of 0.69
"Hideous!" is NOT a toxic comment with a toxicity score of 0.37
"You're fat!" IS a toxic comment with a toxicity score of 0.67
"I hope you trip and die" IS a toxic comment with a toxicity score of 0.85
"Go away!" is NOT a toxic comment with a toxicity score of 0.29
"Your moms a hoe!" IS a toxic comment with a toxicity score of 0.88
"You're better off dead!!" IS a toxic comment with a toxicity score of 0.75
"No one loves you!" IS a toxic comment with a toxicity score of 0.51
"SHUT UP DUMBASS" IS a toxic comment with a toxicity score of 0.90
"I hope you fall" is NOT a toxic comment with a toxicity score of 0.47
"F uk your politeness c 0k su ker" IS a toxic co

My hypothesis was that death threats are more toxic than insults. My hypothesis was proven wrong according to the model that used a threshold of 0.5 on my sample. I was surprised to find out that the results did not show a clear difference in toxicity between the two. Insults such as "You're a complete idiot" and death threats such as "I will kill you" both had toxicity scores above 0.90, which I find to be relatively similar and shocking. You would think that the model would rate "I will kill you" as significantly higher compared to the insult. This made me realize that the model may not be that accurate in reflecting the real world and raises questions about the model. The interesting aspect is that there was no significant difference in toxicity scores between death threats and insults.

I think biases might exist in the data that was used to train this model. The data probably contains biases in the context of the real world and downplays the importance and harm of death threats, which are reflected in the model's performance. The biases might also be related to certain words used and how the model is trained to perceive one word as more severe than another. For instance, the word 'complete' in the insult "You're a complete idiot" might result in higher toxicity than simply saying "I hope you trip and die" because perhaps the model sees the word 'hope' as more positive.

I have several theories as to why the model might also be a poor measure of toxicity. I think the model has limitations in understanding toxicity levels, considering it is not human and cannot process or analyze it the way an ordinary person would. There is also more nuance to this because someone's toxicity might affect someone worse than it would another individual. Toxicity can be subjective. The model is also limited in understanding context and intent, which could also be a factor in scoring the toxicity of an insult versus a death threat.









